internal/controllers/gateway/route_utils.go

```diff
@@ -84,7 +86,62 @@ func getSupportedGatewayForRoute(ctx context.Context, mgrc client.Client, obj cl
 		// if the GatewayClass matches this controller we're all set and this controller
 		// should reconcile this object.
 		if gatewayClass.Spec.ControllerName == ControllerName {
-			gateways = append(gateways, &gateway)
+			allowedNamespaces := make(map[string]interface{})
+			// set true if we find any AllowedRoutes. there may be none, in which case any namespace is permitted
+			filtered := false
+			for _, listener := range gateway.Spec.Listeners {
+				// Only match if the listener type matches the protocol type
+				switch obj.(type) {
+				case *gatewayv1alpha2.HTTPRoute:
+					if !(listener.Protocol == gatewayv1alpha2.HTTPProtocolType || listener.Protocol == gatewayv1alpha2.HTTPSProtocolType) {
+						continue
+					}
+				case *gatewayv1alpha2.TCPRoute:
+					if listener.Protocol != gatewayv1alpha2.TCPProtocolType {
+						continue
+					}
+				case *gatewayv1alpha2.UDPRoute:
+					if listener.Protocol != gatewayv1alpha2.UDPProtocolType {
+						continue
+					}
+				case *gatewayv1alpha2.TLSRoute:
+					if listener.Protocol != gatewayv1alpha2.TLSProtocolType {
+						continue
+					}
+				default:
+					continue
```

Comment:
This is kinda sorta designed around what this should look like in the future--it's a bastardized allowedRoutes.kind implementation (I need to mark that TODO here) that only handles the "if none specified" part of the spec (you get unwanted behavior without it at minimum).

The full-fledged version should continue because it should check against all (there may be multiple listeners that can handle it on a protocol level, but not all actually allow its kind and not all that allow its kind may allow its namespace). The ultimate upshot of those checks is that you get to the end, see no matching gateways, and set the !Accepted NoMatchingAllowedRoutes reason.

As such, I don't think we should throw a different error here since it should eventually go away. `parentRefsForRoute()` already handles unsupported type filtering at a higher level, so we wouldn't be able to hit either variant unless we goof up.